fog/ledger/server/tests/router_connection.rs

```diff
@@ -0,0 +1,908 @@
+            config,
+            enclave,
+            ra_client,
+            Arc::new(RwLock::new(HashMap::new())),
+            ledger.clone(),
+            watcher.clone(),
+            logger.clone(),
+        );
+
+        ledger_server.start();
+
+        // Make unattested ledger client
+        let client = FogUntrustedLedgerGrpcClient::new(
+            client_listen_uri,
+            GRPC_RETRY_CONFIG,
+            grpc_env,
+            logger.clone(),
+        );
+
+        // Try to get a block
+        let queries = [0..1];
+        let result = client.get_blocks(&queries).unwrap();
+        // Check that we got 1 block, as expected
+        assert_eq!(result.blocks.len(), 1);
+        assert_eq!(result.blocks[0].index, 0);
+        assert_eq!(result.blocks[0].outputs.len(), 1);
+        assert_eq!(result.blocks[0].global_txo_count, 1);
+        assert_eq!(
+            result.blocks[0].timestamp_result_code,
+            TimestampResultCode::BlockIndexOutOfBounds as u32
+        );
+        assert_eq!(result.num_blocks, num_blocks);
+        assert_eq!(result.global_txo_count, ledger.num_txos().unwrap());
+
+        // Try to get two blocks
+        let queries = [1..3];
+        let result = client.get_blocks(&queries).unwrap();
+
+        // Check that we got 2 blocks, as expected
+        assert_eq!(result.blocks.len(), 2);
+        assert_eq!(result.blocks[0].index, 1);
+        assert_eq!(result.blocks[0].outputs.len(), 2);
+        assert_eq!(result.blocks[0].global_txo_count, 3);
+        assert_eq!(
+            result.blocks[0].timestamp_result_code,
+            TimestampResultCode::TimestampFound as u32
+        );
+        assert_eq!(result.blocks[1].index, 2);
+        assert_eq!(result.blocks[1].outputs.len(), 3);
+        assert_eq!(result.blocks[1].global_txo_count, 6);
+        assert_eq!(
+            result.blocks[1].timestamp_result_code,
+            TimestampResultCode::TimestampFound as u32
+        );
+        assert_eq!(result.num_blocks, num_blocks);
+        assert_eq!(result.global_txo_count, ledger.num_txos().unwrap());
+    }
+
+    // grpcio detaches all its threads and does not join them :(
+    // we opened a PR here: https://github.com/tikv/grpc-rs/pull/455
+    // in the meantime we can just sleep after grpcio env and all related
+    // objects have been destroyed, and hope that those 6 threads see the
+    // shutdown requests within 1 second.
+    sleep(Duration::from_millis(1000));
+}
+
+// Test that a fog ledger connection is able to check key images by hitting
+// a fog ledger server
+#[test_with_logger]
+fn fog_ledger_untrusted_tx_out_api_test(logger: Logger) {
+    let mut rng = RngType::from_seed([0u8; 32]);
+
+    let alice = AccountKey::random_with_fog(&mut rng);
+    let bob = AccountKey::random_with_fog(&mut rng);
+    let charlie = AccountKey::random_with_fog(&mut rng);
+
+    let recipients = vec![alice.default_subaddress()];
+
```

Comment:
:wrench: similar to above consider removing and moving explicitly into the `add_block_to_ledger()`
```suggestion
```